libp2p/nameresolving/dnsresolver.nim

```diff
@@ -0,0 +1,147 @@
+## Nim-LibP2P
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+{.push raises: [Defect].}
+
+import
+  std/[streams, strutils, sets, sequtils],
+  chronos, chronicles,
+  dnsclientpkg/[protocol, types]
+
+import
+  nameresolver
+
+logScope:
+  topics = "libp2p dnsresolver"
+
+type
+  DnsResolver* = ref object of NameResolver
+    nameServers*: seq[TransportAddress]
+
+proc questionToBuf(address: string, kind: QKind): seq[byte] =
+  try:
+    var
+      header = initHeader()
+      question = initQuestion(address, kind)
+
+      requestStream = header.toStream()
+    question.toStream(requestStream)
+
+    let dataLen = requestStream.getPosition()
+    requestStream.setPosition(0)
+
+    var buf = newSeq[byte](dataLen)
+    discard requestStream.readData(addr buf[0], dataLen)
+    return buf
+  except CatchableError as exc:
+    info "Failed to created DNS buffer", msg = exc.msg
+    return newSeq[byte](0)
+
+proc getDnsResponse(
+  dnsServer: TransportAddress,
+  address: string,
+  kind: QKind): Future[Response] {.async.} =
+
+  var sendBuf = questionToBuf(address, kind)
+
+  if sendBuf.len == 0:
+    raise newException(ValueError, "Incorrect DNS query")
+
+  let receivedDataFuture = newFuture[void]()
+
+  proc datagramDataReceived(transp: DatagramTransport,
+                  raddr: TransportAddress): Future[void] {.async, closure.} =
+      receivedDataFuture.complete()
+
+  let sock =
+    if dnsServer.family == AddressFamily.IPv6:
+      newDatagramTransport6(datagramDataReceived)
+    else:
+      newDatagramTransport(datagramDataReceived)
+
+  try:
+    await sock.sendTo(dnsServer, addr sendBuf[0], sendBuf.len)
+
+    await receivedDataFuture or sleepAsync(5.seconds) #unix default
+
+    if not receivedDataFuture.finished:
+      raise newException(ValueError, "DNS server timeout")
+
+    var
```

Comment:
Should this just be moved into the `datagramDataReceived`? I guess it's a matter of style, but it does look odd in terms of flow...

Reply:
The goal was to have something like that:
```nim
await sock.sendTo(xx)
let result = await sock.recv()
[use result]
```
But chronos udp isn't built like that, so I re-created that as I could. I think this is still clearer than splitting the logic in half (sending and receiving)

Reply:
How so? As far as I can see it's an async callback that triggers on every packet and _should_ be able to manipulate the incoming data? Like I said, this is more a matter of taste, as your approach works, but it does look odd in terms of flow. At least in my case, I would expect that logic pertaining to manipulating the incoming data happens in the callback that triggers for that data? 

Again, matter of style, just an observation.